druid-filtering-ext/src/main/java/bi/deep/filtering/common/IPBoundedRange.java

```diff
@@ -0,0 +1,135 @@
+import java.util.Objects;
+import javax.annotation.Nullable;
+import org.apache.druid.error.InvalidInput;
+import org.apache.druid.java.util.common.IAE;
+
+public class IPBoundedRange {
+    private final String lower;
+    private final String upper;
+    private final boolean lowerOpen;
+    private final boolean upperOpen;
+
+    private final IPAddress lowerIPAddress;
+    private final IPAddress upperIPAddress;
+    private final IPAddress.IPVersion ipVersion;
+
+    @JsonCreator
+    public IPBoundedRange(
+            @JsonProperty("lower") @Nullable String lower,
+            @JsonProperty("upper") @Nullable String upper,
+            @JsonProperty("lowerOpen") @Nullable Boolean lowerOpen,
+            @JsonProperty("upperOpen") @Nullable Boolean upperOpen) {
+
+        if (lower == null && upper == null) {
+            throw InvalidInput.exception("Invalid input, lower and upper cannot be null at the same time");
+        }
+
+        this.lower = lower;
+        this.upper = upper;
+        this.lowerIPAddress = lower != null ? new IPAddressString(lower).getAddress() : null;
+        this.upperIPAddress = upper != null ? new IPAddressString(upper).getAddress() : null;
+        this.lowerOpen = lowerOpen != null && lowerOpen;
+        this.upperOpen = upperOpen != null && upperOpen;
+
+        if (this.lowerIPAddress != null
+                && this.upperIPAddress != null
+                && this.lowerIPAddress.getIPVersion() != this.upperIPAddress.getIPVersion()) {
+            throw new IAE("Both lower and upper bounds must be of the same IP type (IPv4 or IPv6)");
+        }
+
+        this.ipVersion =
+                this.lowerIPAddress != null ? this.lowerIPAddress.getIPVersion() : this.upperIPAddress.getIPVersion();
+    }
+
+    @JsonProperty
+    public String getLower() {
+        return lower;
+    }
+
+    @JsonProperty
+    public String getUpper() {
+        return upper;
+    }
+
+    @JsonProperty
+    public boolean isLowerOpen() {
+        return lowerOpen;
+    }
+
+    @JsonProperty
+    public boolean isUpperOpen() {
+        return upperOpen;
+    }
+
+    public IPAddress getLowerIPAddress() {
+        return lowerIPAddress;
+    }
+
+    public IPAddress getUpperIPAddress() {
+        return upperIPAddress;
+    }
+
+    private boolean matchUpperBound(IPAddress ipValue) {
+        return upperIPAddress == null || upperOpen
+                ? ipValue.compareTo(upperIPAddress) < 0
+                : ipValue.compareTo(upperIPAddress) <= 0;
+    }
+
+    private boolean matchLowerBound(IPAddress ipValue) {
+        return lowerIPAddress == null || lowerOpen
+                ? ipValue.compareTo(lowerIPAddress) > 0
+                : ipValue.compareTo(lowerIPAddress) >= 0;
+    }
+
+    @VisibleForTesting
+    public boolean contain(IPAddress ipAddress, boolean ignoreVersionMismatch) {
+        return ipAddress == null || ipAddress.getIPVersion() != this.ipVersion
+                ? ignoreVersionMismatch
```

Comment:
Should it return true if versionMismatch is ignored?  Dont get the idea

Reply:
Yes, suppose if we wants to apply a range to IPv4 but also wants any IPv6 address.